feat(formly-form): add 'watcher' to fieldGroup

benoror opened the pull request:
originally requested in:
http://stackoverflow.com/questions/34850697/clearing-fieldgroups-fields-when-hidden/34857087?noredirect=1#comment57499232_34857087

benoror commented on src/directives/formly-form.test.js:
```diff
@@ -952,5 +952,31 @@ describe('formly-form', () => {
+          listener,
+          expression,
+        }],
+        key: 'group',
```
@kentcdodds I added a key to the fieldGroup here, otherwise I get:

```
Firefox 43.0.0 (Ubuntu 0.0.0) formly-form field watchers should setup any watchers specified on a fieldgroup FAILED
	expected [Function: compileAndDigest] to not throw an error but 'TypeError: field.key is undefined' was thrown
	AssertionError@/home/benoror/github/angular-formly/node_modules/karma-sinon-chai/node_modules/chai/chai.js:5033:18
	[3]</module.exports/Assertion.prototype.assert@/home/benoror/github/angular-formly/node_modules/karma-sinon-chai/node_modules/chai/chai.js:206:1
	assertThrows@/home/benoror/github/angular-formly/node_modules/karma-sinon-chai/node_modules/chai/chai.js:1629:1
	[10]</module.exports/ctx[name]@/home/benoror/github/angular-formly/node_modules/karma-sinon-chai/node_modules/chai/chai.js:3748:18
	@/home/benoror/github/angular-formly/src/index.test.js:16699:8
```

Not sure why since key is marked as optional

kentcdodds replied:
Yeah, that would indicate to me that anyone using field groups without specifying a `key` on the field group object will get that error when compiling. Could you investigate it further so the key is not required?

benoror replied:
ok, apparently when the expression is an empty string, this commit set it to it's field's [nested]model https://github.com/formly-js/angular-formly/commit/2b69b8e564c6d05699c9e89835fbb8e6b572c64c

I will fix it asap

kentcdodds commented:
@benoror looks like you updated the PR. Looks good to me. Anyone from @formly-js/angular-formly-collaborators or @formly-js/angular-formly-collaborators-read wanna review this and push the merge button?

benoror commented:
I did, I forgot that github doesn't notify PR updates

kamilkisiela commented:
@benoror Are you going to resolve conflicts or should I look it up?

kentcdodds commented:
Sorry, I just merged another PR :cry:

benoror commented:
Np, give me a moment :smile:

benoror commented:
@kamilkisiela Apparently the last PR changed apiCheck for watchers, among other stuff, thus now I get some unit tests failing.

Sorry, this merge is going to take a bit more time, feel free to try it yourself

kentcdodds commented:
Sorry about that @benoror.

kamilkisiela commented:
@benoror Saw that too. Ok :)

kamilkisiela commented:
Ok, done. 

I also fixed one error caused by one of dependencies in karma cfaa247

kentcdodds commented:
:clap:

benoror commented:
@kamilkisiela Great! :smile: 

You reverted some of the changes, for example: [I introduced watcherChecker](https://github.com/formly-js/angular-formly/commit/b2781ef454835268c6c595840243c485ee00e357#diff-07752bb803b21dd1aadc5838eaaf7744R186), then you [reverted it back](https://github.com/formly-js/angular-formly/commit/b2781ef454835268c6c595840243c485ee00e357#diff-07752bb803b21dd1aadc5838eaaf7744R186) :cry: 

Since you also introduced several more commits, would you mind cleaning the git history a little bit?

Thanks!

kamilkisiela commented:
@benoror I just merged your branch and then fix conflicts.

`watcherChecker` is great but it will work only with [`fieldGroup`'s apiChecking](https://github.com/formly-js/angular-formly/blob/c7d8343348a18407b966d9b696bee87150926514/src/providers/formlyApiCheck.js#L187).

[`fieldOptionsApiShape`](https://github.com/formly-js/angular-formly/blob/c7d8343348a18407b966d9b696bee87150926514/src/providers/formlyApiCheck.js#L116) has one more option `runFieldExpressions`. 

I think that ApiCheck has no "extending" function so this `watcherChecker` would be used only in one place.

Correct me if I'm wrong.

About mess in git history :) I'm going to work on it tomorrow.

benoror commented:
maybe the new `runFieldExpressions` field should be in both places?

kamilkisiela commented:
@kentcdodds What do you think?

kentcdodds commented:
Ah, yes, probably should.

kamilkisiela commented:
@benoror Done :)

benoror commented:
@kamilkisiela thanks a lot! :tada:
